src/bit-fields-macros/src/lib.rs

```diff
@@ -0,0 +1,1065 @@
+        #[doc=#header]
+        ///
+        /// ## Layout
+        ///
+        /// <table>
+        #struct_doc_table_layout
+        /// </table>
+        #[derive(Clone)]
+        #[repr(C)]
+        pub struct #struct_name {
+            pub data: #struct_data_type,
+            pub bits: (#struct_bits),
+            #struct_member_fields
+        }
+        #serde
+        impl std::cmp::PartialEq<#struct_data_type> for #struct_name {
+            fn eq(&self,other:&#struct_data_type) -> bool {
+                self.data == *other
+            }
+        }
+        impl std::cmp::PartialEq for #struct_name {
+            fn eq(&self,other: &Self) -> bool {
+                self.data == other.data
+            }
+        }
+        impl std::cmp::Eq for #struct_name { }
+        /// We cannot derive [`std::fmt::Debug`] as `self.bits` may have too many elements.
+        impl std::fmt::Debug for #struct_name {
+            fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+                f.debug_tuple(#struct_name_str)
+                    .field(&self.data)
+                    .finish()
+            }
+        }
+        #display_impl
+        /// We cannot derive [`std::default::Default`] as `self.bits` may have too many elements.
+        impl std::default::Default for #struct_name {
+            fn default() -> Self {
+                Self::from(0)
+            }
+        }
+        impl std::fmt::Binary for #struct_name {
+            fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+                std::fmt::Binary::fmt(&self.data, f)
+            }
+        }
+        #flag_set
+        #field_map
+        #set_map_conversions
+        /// Constructs `self` with the data type.
+        impl std::convert::From<#struct_data_type> for #struct_name {
+            fn from(data: #struct_data_type) -> Self {
+                Self {
+                    data,
+                    bits: (#struct_new_bits),
+                    #struct_member_fields_initialization
+                }
+            }
+        }
+        /// Constructs the data type from `self`.
+        impl std::convert::From<#struct_name> for #struct_data_type {
+            fn from(bit_field: #struct_name) -> Self {
+                bit_field.data
+            }
+        }
+        impl #struct_name {
+            #cmp_flags
+            #set_theory
+            #index_fn
+        }
+        #bit_index
+    };
+    // eprintln!("\n------------------------------------------------------\
+    // n------------------------------------------------------\
+    // n------------------------------------------------------\nlayout:\n{}\
+    // n------------------------------------------------------\
+    // n------------------------------------------------------\
+    // n------------------------------------------------------\n", layout); panic!("bruh");
+    // "fn answer() -> u32 { 42 }".parse().unwrap()
+    proc_macro::TokenStream::from(layout)
+}
```

Comment:
Since we're using `proc_macro2` we can also include unit tests for the proc macro. We need to add unit tests for it.

Reply:
THe unit tests are under `bit-fields/lib.rs` the macro in `bit-fields-macros`does not work without types defined in `bit-fields`.

Reply:
I've added units tests under `utils.rs` for `MultiLineString` and `SliceIter`. All the functions of `DataType` are simple and can be evaluated at compile time, I am not sure it is necessary to unit test them.

`proc_macro::Span`s (and `proc_macro2::Span`s) cannot be constructed outside proc-macros, as a result `BitFieldBuilder` cannot be directly constructed in a unit test. Given this I do not think we can write direct unit tests for `BitFieldBuilder`.